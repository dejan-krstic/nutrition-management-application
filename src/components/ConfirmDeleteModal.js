import React from "react";
import { Modal } from "antd";

const ConfirmDeleteModal = (props) => (
  <>
    <Modal
      title={props.title}
      visible={props.visible}
      onOk={() => props.onOk(props.id)}
      onCancel={props.omCancel}
    >
      {props.children}
    </Modal>
  </>
);

export default ConfirmDeleteModal;
