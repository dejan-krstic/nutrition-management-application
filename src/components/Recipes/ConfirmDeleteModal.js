import { Modal } from "antd";

const ConfirmDeleteModal = (props) => (
  <>
    <Modal
      title="Do you want to delete this recipe"
      visible={props.visible}
      onOk={props.handleOk}
      onCancel={props.handleCancel}
    />
  </>
);

export default ConfirmDeleteModal;
