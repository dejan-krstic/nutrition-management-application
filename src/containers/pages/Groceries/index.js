import { Col, Row, Spin } from "antd";
import { connect } from "react-redux";
import {
  groceriesSelector,
  isLoadingGroceriesSelector,
  areLoadedGroceriesSelector,
} from "../../../selectors";
import ItemList from "../../../components/ItemList";
import { GROCERIES } from "../../../constants";
import { setNewAmount, getGroceries } from "../../../actions";
import { useCallback, useEffect } from "react";

const Groceries = (props) => {
  const handleAmountChange = useCallback(
    (name, value) => {
      props.setNewAmount(name, value);
    },
    [props.setNewAmount],
  );

  useEffect(() => {
    if (!props.isLoadingGroceries && !props.groceriesLoaded) {
      props.getGroceries();
    }
  }, [props.isLoadingGroceries, props.groceriesLoaded]);

  return (
    <Row justify="space-around">
      <Col xs={22} sm={20} md={16} lg={12} xl={10} xxl={7}>
        {props.isLoadingGroceries ? (
          <Row justify="center" style={{ marginTop: "30vh" }}>
            <Spin size="large" />
          </Row>
        ) : (
          <ItemList
            {...props}
            header={GROCERIES}
            handleAmountChange={handleAmountChange}
          />
        )}
      </Col>
    </Row>
  );
};

const mapStateToProps = (state) => ({
  items: groceriesSelector(state),
  isLoadingGroceries: isLoadingGroceriesSelector(state),
  groceriesLoaded: areLoadedGroceriesSelector(state),
});
export default connect(mapStateToProps, { setNewAmount, getGroceries })(
  Groceries,
);
