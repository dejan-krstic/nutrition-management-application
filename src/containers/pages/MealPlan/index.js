import { Col, Row, Spin } from "antd";
import WhatToEatDayList from "../../../components/WhatToEat/WhatToEatDayList";
import { connect } from "react-redux";
import {
  mealPlanSelector,
  recipesSelector,
  areLoadedMealsSelector,
  areLoadedRecipesSelector,
  isLoadingRecipesSelector,
  isLoadingMealsSelector,
  mealsSelector,
} from "../../../selectors";
import { choseMealOnDate, getRecipes, getMeals } from "../../../actions";
import WhatToEatModal from "../../../components/WhatToEat/WhatToEatModal";
import { useEffect, useState } from "react";

const MealPlan = (props) => {
  const [showModal, showWhatToEatModal] = useState(false);
  const [mealData, setMealData] = useState({});

  const showWhatToEatModalHandler = (meal) => {
    setMealData(meal);
    showWhatToEatModal(true);
  };

  const closeWhatToEatModalHandler = () => showWhatToEatModal(false);

  const handleChoose = (id) => {
    props.choseMealOnDate(props.meals, mealData.date, mealData.type, id);
    closeWhatToEatModalHandler();
  };

  useEffect(() => {
    if (!props.isLoadingRecipes && !props.recipesLoaded) {
      props.getRecipes();
    }
  }, [props.isLoadingRecipes, props.recipesLoaded]);

  useEffect(() => {
    if (!props.isLoadingMeals && !props.mealsLoaded) {
      props.getMeals();
    }
  }, [props.isLoadingMeals, props.mealsLoaded]);

  return (
    <>
      <WhatToEatModal
        visible={showModal}
        handleChose={handleChoose}
        handleClose={closeWhatToEatModalHandler}
        items={props.recipes.filter((r) => r.category === mealData.type)}
      />
      <Row justify="center">
        <Col xs={23} sm={22} md={21} xl={20} xxl={17}>
          {props.isLoadingRecipes || props.isLoadingMeals ? (
            <Row justify="center" style={{ marginTop: "30vh" }}>
              <Spin size="large" />
            </Row>
          ) : (
            <WhatToEatDayList
              {...props}
              addMealModalHandler={showWhatToEatModalHandler}
            />
          )}
        </Col>
      </Row>
    </>
  );
};

const mapStateToProps = (state) => ({
  mealPlan: mealPlanSelector(state),
  meals: mealsSelector(state),
  recipes: recipesSelector(state),
  isLoadingRecipes: isLoadingRecipesSelector(state),
  isLoadingMeals: isLoadingMealsSelector(state),
  recipesLoaded: areLoadedRecipesSelector(state),
  mealsLoaded: areLoadedMealsSelector(state),
});

export default connect(mapStateToProps, {
  choseMealOnDate,
  getMeals,
  getRecipes,
})(MealPlan);
