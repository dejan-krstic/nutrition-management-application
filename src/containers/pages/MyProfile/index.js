import { connect } from "react-redux";
import AddNewRecipe from "../../../components/AddNewRecipe";
import { Typography } from "antd";
import { addNewRecipe } from "../../../actions";
import { useCallback } from "react";
import { locationSelector } from "../../../selectors";

const MyProfile = ({ addNewRecipe, location }) => {
  const addNewRecipeHandler = useCallback(
    (recipe) => {
      addNewRecipe(recipe);
    },
    [addNewRecipe],
  );

  return (
    <>
      <Typography.Title
        level={1}
        style={{ textAlign: "center", color: "#666" }}
      >
        ADD NEW RECIPE
      </Typography.Title>
      <AddNewRecipe
        onAddNewRecipe={addNewRecipeHandler}
        recipe={location && location.state && JSON.parse(location.state)}
      />
      ;
    </>
  );
};

const mapStateToProps = (state) => ({
  location: locationSelector(state),
});

export default connect(mapStateToProps, { addNewRecipe })(MyProfile);
