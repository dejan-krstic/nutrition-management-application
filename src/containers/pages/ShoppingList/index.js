import { Col, Row, Typography, Slider, InputNumber } from "antd";
import { connect } from "react-redux";
import {
  shoppingListSelector,
  daysToProvideSelector,
} from "../../../selectors";
import ItemList from "../../../components/ItemList";
import { SHOPPING_LIST } from "../../../constants";
import { useCallback } from "react";
import { setDaysToProvide } from "../../../actions";

const ShoppingList = (props) => {
  const handleChangeDaysToProvide = useCallback(
    (daysToProvide) => {
      props.setDaysToProvide(daysToProvide);
    },
    [setDaysToProvide],
  );

  return (
    <Row justify="center">
      <Col xs={22} sm={20} md={16} lg={12} xl={10} xxl={7}>
        <Row justify="center">
          <Typography.Title
            level={4}
            style={{ textAlign: "center", color: "#666" }}
          >
            SET NUMBER OF DAYS TO PROVIDE FOR:
          </Typography.Title>
        </Row>
        <Row style={{ marginBottom: "1rem" }} justify="center">
          <Col xs={15} sm={17}>
            <Slider
              min={1}
              max={30}
              onChange={handleChangeDaysToProvide}
              value={
                typeof props.daysToProvide === "number"
                  ? props.daysToProvide
                  : 1
              }
            />
          </Col>
          <Col xs={2} sm={3}>
            <InputNumber
              min={1}
              max={30}
              style={{ margin: "0 1rem", fontSize: "1.5rem" }}
              value={props.daysToProvide}
              onChange={handleChangeDaysToProvide}
            />
          </Col>
        </Row>
        <Row>
          <ItemList
            {...props}
            header={SHOPPING_LIST}
            style={{ width: "100%" }}
          />
        </Row>
      </Col>
    </Row>
  );
};

const mapStateToProps = (state) => ({
  items: shoppingListSelector(state),
  daysToProvide: daysToProvideSelector(state),
});

export default connect(mapStateToProps, { setDaysToProvide })(ShoppingList);
