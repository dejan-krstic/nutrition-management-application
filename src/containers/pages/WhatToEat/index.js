import React from "react";

import RecipeList from "../../../components/Recipes/RecipeList";
import { Col, Row } from "antd";
import { multiplyRecipes } from "../../../constants";

const WhatToEat = () => {
  return (
    <Row>
      <Col xs={1} md={4} />
      <Col xs={22} md={16}>
        <RecipeList items={multiplyRecipes(25)} />
      </Col>
      <Col xs={1} md={4} />
    </Row>
  );
};

export default WhatToEat;
