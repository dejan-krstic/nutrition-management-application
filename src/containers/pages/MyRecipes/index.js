import React, { useEffect, useState } from "react";
import RecipeList from "../../../components/Recipes/RecipeList";
import { Col, Row, Empty, Button, Spin } from "antd";
import {
  multiplyRecipes,
  CONFIRM_DELETE_RECIPE_SUCCESS,
  NAVIGATION_LINKS,
} from "../../../constants";
import { connect } from "react-redux";
import ConfirmDeleteModal from "../../../components/ConfirmDeleteModal";
import RecipeModal from "../../../components/Recipes/RecipeModal";
import {
  recipesSelector,
  isLoadingRecipesSelector,
  areLoadedRecipesSelector,
} from "../../../selectors";
import { deleteRecipe, getRecipes } from "../../../actions";
import navigationHook from "../../../hooks/navigation";

export const MyRecipes = ({
  recipes,
  isLoadingRecipes,
  recipesLoaded,
  deleteRecipe,
  getRecipes,
}) => {
  const [showRecipeModal, setOpenRecipeModal] = useState(false);
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [recipeModalData, setRecipeModalData] = useState({});
  const { historyPush } = navigationHook();

  const showRecipeModalHandler = (recipe) => {
    setRecipeModalData(recipe);
    setOpenRecipeModal(true);
  };

  const closeRecipeModalHandler = () => setOpenRecipeModal(false);

  const handleDelete = () => {
    setShowConfirmModal(true);
  };

  const cancelDelete = () => {
    setShowConfirmModal(false);
  };

  const confirmDeleteHandler = (id) => {
    deleteRecipe(id);
    setShowConfirmModal(false);
    closeRecipeModalHandler();
  };

  const handleCreateNew = () => {
    historyPush(NAVIGATION_LINKS.myProfile.path);
  };

  const handleRecipeEdit = () => {
    historyPush(
      NAVIGATION_LINKS.myProfile.path,
      JSON.stringify(recipeModalData),
    );
  };

  useEffect(() => {
    if (!isLoadingRecipes && !recipesLoaded) {
      getRecipes();
    }
  }, [isLoadingRecipes, recipesLoaded]);

  return (
    <>
      {!recipes.length && !isLoadingRecipes && (
        <Empty
          image="https://gw.alipayobjects.com/zos/antfincdn/ZHrcdLPrvN/empty.svg"
          imageStyle={{
            height: 60,
          }}
          description={<span>No recipes exist, maybe create one?</span>}
        >
          <Button type="primary" onClick={handleCreateNew}>
            Create Now
          </Button>
        </Empty>
      )}
      <Row justify="space-around">
        <Col xs={23} xxl={16}>
          {isLoadingRecipes ? (
            <Row justify="center" style={{ marginTop: "30vh" }}>
              <Spin size="large" />
            </Row>
          ) : (
            <RecipeList
              items={multiplyRecipes(3, recipes)}
              showRecipeModalHandler={showRecipeModalHandler}
            />
          )}
        </Col>
      </Row>
      <ConfirmDeleteModal
        title={CONFIRM_DELETE_RECIPE_SUCCESS}
        id={recipeModalData.id}
        visible={showConfirmModal}
        onCancel={cancelDelete}
        onOk={confirmDeleteHandler}
      >
        {recipeModalData.title}
      </ConfirmDeleteModal>
      <RecipeModal
        {...recipeModalData}
        visible={showRecipeModal}
        onDelete={handleDelete}
        onEdit={handleRecipeEdit}
        onClose={closeRecipeModalHandler}
      />
    </>
  );
};

const mapStateToProps = (state) => ({
  recipes: recipesSelector(state),
  isLoadingRecipes: isLoadingRecipesSelector(state),
  recipesLoaded: areLoadedRecipesSelector(state),
});

export default connect(mapStateToProps, { deleteRecipe, getRecipes })(
  MyRecipes,
);
