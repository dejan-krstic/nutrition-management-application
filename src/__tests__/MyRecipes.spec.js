import "jsdom-global/register";
import React from "react";
import { shallow } from "enzyme";
import { MyRecipes } from "../containers/pages/MyRecipes";
import { Empty, Spin } from "antd";
import ConfirmDeleteModal from "../components/ConfirmDeleteModal";

jest.mock("../actions", () => ({
  choseMealOnDate: jest.fn(),
  getRecipes: jest.fn(),
  getMeals: jest.fn(),
  getGroceries: jest.fn(),
  addNewRecipe: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

beforeEach(() => {
  jest.resetAllMocks();
});

beforeAll(() => {
  Object.defineProperty(window, "matchMedia", {
    value: jest.fn(() => {
      return {
        matches: true,
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
      };
    }),
  });
});

describe("MyRecipes", () => {
  const defaultProps = {
    recipes: [],
    isLoadingRecipes: 0,
    recipesLoaded: false,
    deleteRecipe: jest.fn(),
    getRecipes: jest.fn(),
  };

  it("should match snapshot with Empty page", () => {
    const wrapper = shallow(<MyRecipes {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with loader active", () => {
    const props = { ...defaultProps, isLoadingRecipes: 1 };
    const wrapper = shallow(<MyRecipes {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with recipes present", () => {
    const props = { ...defaultProps, isLoadingRecipes: 0, recipes: [{}] };
    const wrapper = shallow(<MyRecipes {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should load Empty when there are no recipes", () => {
    const wrapper = shallow(<MyRecipes {...defaultProps} />);
    const empty = wrapper.find(Empty);

    expect(empty).toHaveLength(1);
  });

  it("should not load Empty when there are recipes or they are loading", () => {
    const props = { ...defaultProps, isLoadingRecipes: 1, recipes: [1] };
    const wrapper = shallow(<MyRecipes {...props} />);
    const empty = wrapper.find(Empty);

    expect(empty).toHaveLength(0);
  });

  it("should find ConfirmDeleteModal and delete recipe with selected id", () => {
    const wrapper = shallow(<MyRecipes {...defaultProps} />);
    const recipeId = 5;

    const cdm = wrapper.find(ConfirmDeleteModal);
    cdm.simulate("ok", recipeId);

    expect(defaultProps.deleteRecipe).toHaveBeenCalledWith(recipeId);
  });

  it("should load spinner while recipes are loading", () => {
    const props = { ...defaultProps, isLoadingRecipes: 1 };
    const wrapper = shallow(<MyRecipes {...props} />);

    const spin = wrapper.find(Spin);
    expect(spin).toHaveLength(1);
  });

  it("shouldn't render spinner when recipes are not loaded", () => {
    const wrapper = shallow(<MyRecipes {...defaultProps} />);

    const spin = wrapper.find(Spin);
    expect(spin).toHaveLength(0);
  });
});
