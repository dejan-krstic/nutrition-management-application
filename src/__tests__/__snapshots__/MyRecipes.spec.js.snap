// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyRecipes should match snapshot with Empty page 1`] = `ShallowWrapper {}`;

exports[`MyRecipes should match snapshot with loader active 1`] = `ShallowWrapper {}`;

exports[`MyRecipes should match snapshot with recipes present 1`] = `ShallowWrapper {}`;
